In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
%matplotlib inline

In [62]:
os.rename("WELLCOME_APCspend2013_forThinkful.csv", "APCspend2013")

FileNotFoundError: [WinError 2] The system cannot find the file specified: 'WELLCOME_APCspend2013_forThinkful.csv' -> 'APCspend2013'

In [23]:
df = pd.read_csv("APCspend2013", encoding = "ISO-8859-1")

In [24]:
df.head(n=20)

PMID/PMCID  \
0                            NaN   
1                     PMC3679557   
2          23043264  PMC3506128    
3            23438330 PMC3646402   
4           23438216 PMC3601604    
5                     PMC3579457   
6                     PMC3709265   
7           23057412 PMC3495574    
8              PMCID: PMC3780468   
9              PMCID: PMC3621575   
10             PMCID: PMC3739413   
11             PMCID: PMC3530961   
12             PMCID: PMC3624797   
13                    PMC3413243   
14                    PMC3694353   
15                    PMC3572711   
16                      22610094   
17             PMCID: PMC3586974   
18  23455506  PMCID: PMC3607399    
19    PMID: 24015914 PMC3833349    

                                       Publisher  \
0                                            CUP   
1                                            ACS   
2                                            ACS   
3                                            ACS   
4                                            ACS   
5                                            ACS   
6                                            ACS   
7                                            ACS   
8   ACS (Amercian Chemical Society) Publications   
9   ACS (Amercian Chemical Society) Publications   
10  ACS (Amercian Chemical Society) Publications   
11  ACS (Amercian Chemical Society) Publications   
12  ACS (Amercian Chemical Society) Publications   
13                              ACS Publications   
14                              ACS Publications   
15                                 AGA Institute   
16                                         AMBSB   
17         American Association of Immunologists   
18         American Association of Immunologists   
19                     American Chemical Society   

                                   Journal title  \
0                         Psychological Medicine   
1                              Biomacromolecules   
2                                     J Med Chem   
3                                     J Med Chem   
4                                     J Org Chem   
5                 Journal of Medicinal Chemistry   
6                   Journal of Proteome Research   
7                                      Mol Pharm   
8                           ACS Chemical Biology   
9                           ACS Chemical Biology   
10  Journal of Chemical Information and Modeling   
11                Journal of Medicinal Chemistry   
12                Journal of Medicinal Chemistry   
13                                  Biochemistry   
14                Journal of Medicinal Chemistry   
15                              Gastroenterology   
16               Journal of Biological Chemistry   
17                         Journal of Immunology   
18                         Journal of Immunology   
19                          ACS Chemical Biology   

                                        Article title  \
0   Reduced parahippocampal cortical thickness in ...   
1   Structural characterization of a Model Gram-ne...   
2   Fumaroylamino-4,5-epoxymorphinans and related ...   
3   Orvinols with mixed kappa/mu opioid receptor a...   
4   Regioselective opening of myo-inositol orthoes...   
5   Comparative Structural and Functional Studies ...   
6   Mapping Proteolytic Processing in the Secretom...   
7   Quantitative silencing of EGFP reporter gene b...   
8   A Novel Allosteric Inhibitor of the Uridine Di...   
9   Chemical proteomic analysis reveals the drugab...   
10  Locating Sweet Spots for Screening Hits and Ev...   
11  Synthesis and evaluation of ?-thymidine analog...   
12  Discovery and structure-activity relationships...   
13  Monomeric 14-3-3 has a chaperone-like activity...   
14  Synthesis and in vitro and in vivo characteriz...   
15  Much of the genetic risk of colorectal cancer ...   
16                Annexin-1 interaction with FPR2/ALX   
17  Sialoadhesin ligand expression identifies a su...

**Directions**: Determine the five most common journals and the total articles for each. Next, calculate the mean, median, and standard deviation of the open-access cost per article for each journal. For a real bonus round, identify the open access prices paid by subject area.

In [25]:
df["Journal title"] = df["Journal title"].str.lower()

In [39]:
df["Journal title"] = df["Journal title"].str.replace("bio$", "biological").str.replace("plosone", "plos one").str.replace("j med chem", "journal of medicinal chemistry").str.replace("j org chem", "journal of organic chemistry").str.replace("neuroscience$", "neurosciences").str.replace("chem$", "chemistry")

In [40]:
df["Journal title"].value_counts()

plos one                                                               199
journal of biological chemistry                                         53
neuroimage                                                              29
plos genetics                                                           24
plos pathogens                                                          24
nucleic acids research                                                  23
plos neglected tropical diseases                                        20
proceedings of the national academy of sciences                         20
human molecular genetics                                                19
nature communications                                                   19
brain                                                                   14
bmc public health                                                       14
movement disorders                                                      13
journal of neurosciences 

ANSWER: The five most common journals are...
1. Plos One (199 articles)
2. Journal of Biological Chemistry (53 articles)
3. Neuroimage (29 articles)
4. Plos Pathogens (24 articles)
5. Plos Genetics (24 articles)

In [19]:
df.dtypes

PMID/PMCID                                             object
Publisher                                              object
Journal title                                          object
Article title                                          object
COST (£) charged to Wellcome (inc VAT when charged)    object
dtype: object

def multi(x):
    if "$" in df["COST (£) charged to Wellcome (inc VAT when charged)"]:
        return 1.00
    if "£" in df["COST (£) charged to Wellcome (inc VAT when charged)"]:
        return 1.32

df["Multiplier"] = df.apply(multi, axis=1)    

In [43]:
df['Multiplier'] = np.where(df["COST (£) charged to Wellcome (inc VAT when charged)"].str.contains("£"), 1.32, 
                            np.where(df["COST (£) charged to Wellcome (inc VAT when charged)"].str.contains("$"), 1.0, 0.0))



In [44]:
df.head()

PMID/PMCID Publisher                   Journal title  \
0                    NaN       CUP          psychological medicine   
1             PMC3679557       ACS               biomacromolecules   
2  23043264  PMC3506128        ACS  journal of medicinal chemistry   
3    23438330 PMC3646402       ACS  journal of medicinal chemistry   
4   23438216 PMC3601604        ACS    journal of organic chemistry   

                                       Article title  \
0  Reduced parahippocampal cortical thickness in ...   
1  Structural characterization of a Model Gram-ne...   
2  Fumaroylamino-4,5-epoxymorphinans and related ...   
3  Orvinols with mixed kappa/mu opioid receptor a...   
4  Regioselective opening of myo-inositol orthoes...   

  COST (£) charged to Wellcome (inc VAT when charged)  Multiplier  
0                                              £0.00         1.32  
1                                           £2381.04         1.32  
2                                            £642.56         1.32  
3                                            £669.64         1.32  
4                                            £685.88         1.32

In [45]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = df["COST (£) charged to Wellcome (inc VAT when charged)"].str.replace("£", "")

In [46]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = df["COST (£) charged to Wellcome (inc VAT when charged)"].str.replace("$", "")

In [47]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = pd.to_numeric(df["COST (£) charged to Wellcome (inc VAT when charged)"])

In [48]:
df["Cost in Dollars"] = df["COST (£) charged to Wellcome (inc VAT when charged)"] * df["Multiplier"]

In [49]:
df.head()

PMID/PMCID Publisher                   Journal title  \
0                    NaN       CUP          psychological medicine   
1             PMC3679557       ACS               biomacromolecules   
2  23043264  PMC3506128        ACS  journal of medicinal chemistry   
3    23438330 PMC3646402       ACS  journal of medicinal chemistry   
4   23438216 PMC3601604        ACS    journal of organic chemistry   

                                       Article title  \
0  Reduced parahippocampal cortical thickness in ...   
1  Structural characterization of a Model Gram-ne...   
2  Fumaroylamino-4,5-epoxymorphinans and related ...   
3  Orvinols with mixed kappa/mu opioid receptor a...   
4  Regioselective opening of myo-inositol orthoes...   

   COST (£) charged to Wellcome (inc VAT when charged)  Multiplier  \
0                                               0.00          1.32   
1                                            2381.04          1.32   
2                                             642.56          1.32   
3                                             669.64          1.32   
4                                             685.88          1.32   

   Cost in Dollars  
0           0.0000  
1        3142.9728  
2         848.1792  
3         883.9248  
4         905.3616

In [52]:
df["Cost in Dollars"].mean()

31766.058265726464

In [56]:
df["Cost in Dollars"].mode()

0    2692.8
dtype: float64

In [54]:
df["Cost in Dollars"].median()

2486.4972000000002

In [53]:
df["Cost in Dollars"].std()

193856.51704674127

In [44]:
df.groupby("Journal title").describe()

COST (£) charged to Wellcome (inc VAT when charged)  \
                                                                                                 count   
Journal title                                                                                            
academy of nutrition and dietetics                                                                1.0    
acs chemical biology                                                                              5.0    
acs chemical neuroscience                                                                         1.0    
acs nano                                                                                          2.0    
acta crystallographica section d,  biological c...                                                1.0    
acta crystallographica section d: biological cr...                                                1.0    
acta crystallographica section f: structural bi...                                                2.0    
acta crystallographica, section d                                                                 1.0    
acta crystallography d                                                                            1.0    
acta d                                                                                            1.0    
acta dermato venereologica                                                                        1.0    
acta diabetologica                                                                                1.0    
acta f                                                                                            1.0    
acta neuropathol                                                                                  1.0    
acta neuropathologica                                                                             3.0    
acta opthalmologica                                                                               1.0    
acta physiol                                                                                      1.0    
addiction                                                                                         2.0    
advances in experimental medicine and biology                                                     4.0    
age                                                                                               2.0    
age and ageing                                                                                    1.0    
ageing & society                                                                                  1.0    
aging cell                                                                                        1.0    
aids                                                                                              3.0    
aids behav                                                                                        1.0    
aids care                                                                                         4.0    
aids journal                                                                                      1.0    
aids research and therapy                                                                         1.0    
aids uk                                                                                           2.0    
alcohol and alcoholism                                                                            1.0    
...                                                                                               ...    
traffic                                                                                           6.0    
transactions of the royal society of tropical m...                                                1.0    
translational psychiatry                                                                          2.0    
trends in cognitive science                                                                       1.0    
trends in genetics     

The above chart shows (in the "mean" column) the price per article for a given journal, which is not exactly by subject.

In [67]:
df.groupby("Publisher").describe()

Article title         \
                                                           count unique   
Publisher                                                                 
ACS                                                            7      7   
ACS (Amercian Chemical Society) Publications                   5      5   
ACS Publications                                               2      2   
AGA Institute                                                  1      1   
AMBSB                                                          1      1   
AMERICAN CHEMICAL SOCIETY                                      1      1   
ASBMB                                                         46     46   
ASBMB Cadmus                                                   1      1   
ASBMB/Cadmus                                                   1      1   
ASBMB/Cenveo Publisher Services                                1      1   
ASBMC /CENVEO                                                  1      1   
ASM                                                            2      2   
ASM (American Society for Microbiology)                        1      1   
American Association of Immunologists                          2      2   
American Chemical Society                                     14     14   
American Chemical Society Publications                         1      1   
American College of Chest Physicians                           1      1   
American Physiological Society                                 7      7   
American Psychiatric Association                               1      1   
American Psychiatric Publishing                                1      1   
American Psychological Association                            13     13   
American Psychological Association                             1      1   
American Public Health Association                             1      1   
American Soc for Biochemistry and Molecular Bio...             1      1   
American Society for Biochemistry and Molecular...             1      1   
American Society for Biochemistry and Molecular...            18     18   
American Society for Biochemistry and Molecular...             1      1   
American Society for Investigative Pathology                   1      1   
American Society for Microbiology                             16     16   
American Society for Microbiology                              1      1   
...                                                          ...    ...   
The Royal College of Psychiatrists                             2      2   
The Royal Society                                              6      6   
The Sheridan Press                                             3      3   
The Sheridan Press                                             1      1   
The company of Biolgists                                       2      2   
The company of Biologists                                      1      1   
The royal Society                                              1      1   
Transcript Verlag                                              1      1   
University of the Basque Country Press                         1      1   
Wiley                                                        136    136   
Wiley                                                         12     12   
Wiley & Son                                                    1      1   
Wiley Blackwell                                                4      4   
Wiley Online Library                                           1      1   
Wiley Subscription Services                                    8      8   
Wiley Subscription Services  Inc.                             13     13   
Wiley Subscription Services Inc                                2      2   
Wiley Subscription Serviices Inc                               1      1   
Wiley VCH                                                      1      1   
Wiley-Blackwell                                               56     56   
W

The above chart groups the articles by publisher, which probably is closer to grouping by subject.

In [68]:
import re

journals = df["Journal title"]

subject = ["biology"]

for journal in journals:
    if re.search(journal, subject):
        print("found a match")
#df["Journal title"].apply(lambda x: re.findall("biology", x)
                          
#example = ["veterinary studies", "journal of biology", "biology now"]

#re.findall("biology", example)

TypeError: expected string or bytes-like object

In [69]:
import re

In [78]:
df["Journal title"].apply(lambda x: re.findall("biology", x))

TypeError: expected string or bytes-like object

In [83]:
df["Subject"] = ""

In [84]:
df.head()

PMID/PMCID Publisher           Journal title  \
0                    NaN       CUP  psychological medicine   
1             PMC3679557       ACS       biomacromolecules   
2  23043264  PMC3506128        ACS              j med chem   
3    23438330 PMC3646402       ACS              j med chem   
4   23438216 PMC3601604        ACS              j org chem   

                                       Article title  \
0  Reduced parahippocampal cortical thickness in ...   
1  Structural characterization of a Model Gram-ne...   
2  Fumaroylamino-4,5-epoxymorphinans and related ...   
3  Orvinols with mixed kappa/mu opioid receptor a...   
4  Regioselective opening of myo-inositol orthoes...   

  COST (£) charged to Wellcome (inc VAT when charged) Subject  
0                                              £0.00           
1                                           £2381.04           
2                                            £642.56           
3                                            £669.64           
4                                            £685.88

In [98]:
subj = "biology"
journal = "biology rocks"
new_column = []

if subj in df["Journal title"]:
    df["Subject"] = "biology"


In [107]:
print(df["Journal title"])

0                                  psychological medicine
1                                       biomacromolecules
2                                              j med chem
3                                              j med chem
4                                              j org chem
5                          journal of medicinal chemistry
6                            journal of proteome research
7                                               mol pharm
8                                    acs chemical biology
9                                    acs chemical biology
10           journal of chemical information and modeling
11                         journal of medicinal chemistry
12                         journal of medicinal chemistry
13                                           biochemistry
14                         journal of medicinal chemistry
15                                       gastroenterology
16                        journal of biological chemistry
17            

In [78]:
def subject(c):
    if ("biology" in str(c["Journal title"])) or ("blood" in str(c["Journal title"])):
        return "biology"
    elif "chem" in str(c["Journal title"]):
        return "chemistry"
    elif "psych" in str(c["Journal title"]):
        return "psychology"
    elif ("immunology" in str(c["Journal title"])) or ("infection" in str(c["Journal title"])):
        return "immunology"
    elif "physiology" in str(c["Journal title"]):
        return "physiology"
    elif "aids" in str(c["Journal title"]):
        return "AIDS"
    else:
        return "undefined"
    
df["Subject"] = df.apply(subject, axis=1)
    

In [79]:
if "chem" in "j med chem":
    print("yup")

yup


In [80]:
df.tail(n=10)

PMID/PMCID                      Publisher  \
2117                       PMC3773237                  Wolters Kluwer   
2118                       PMC3815011                  Wolters Kluwer   
2119                        PMC3805356                 Wolters Kluwer   
2120                             Epub                  Wolters Kluwer   
2121                        PMC3707567                 Wolters Kluwer   
2122                           2901593          Wolters Kluwer Health   
2123                           3748854         Wolters Kluwer Health    
2124                           3785148         Wolters Kluwer Health    
2125          PMCID:\n    PMC3647051\n  Wolters Kluwer N.V./Lippinott   
2126  PMID: 23846567  (Epub July 2013)                Wolters Kluwers   

                                          Journal title  \
2117                                            aids uk   
2118                                            aids uk   
2119  jaids journal of acquired immune deficiency sy...   
2120  journal of acquired immune deficiency syndroms...   
2121                                    journal of aids   
2122                               circulation research   
2123                                               aids   
2124                               pediatr infect dis j   
2125                                               aids   
2126    journal of acquired immune deficiency syndromes   

                                          Article title  \
2117   Sexual behaviour in a rural high HIV prevalence    
2118  HIV incidence and survival from age-specific s...   
2119  Accelerated biological ageing in HIV-infected ...   
2120  Reduction in early mortality on antiretroviral...   
2121  Risk factors for seropositivity to Kaposi sarc...   
2122  Mechanistic Links Between Na+ Channel (SCN5A) ...   
2123  Evaluation of an empiric risk screening score ...   
2124  Topical umbilical cord care for prevention of ...   
2125  Grassroots Community Organisations' Contributi...   
2126  A novel community health worker tool outperfor...   

      COST (£) charged to Wellcome (inc VAT when charged)  Multiplier  \
2117                                            1836.92          1.32   
2118                                            1836.92          1.32   
2119                                            1375.80          1.32   
2120                                            1836.92          1.32   
2121                                            2009.65          1.32   
2122                                            1334.15          1.32   
2123                                            1834.77          1.32   
2124                                            1834.77          1.32   
2125                                            2374.52          1.32   
2126                                            2034.75          1.32   

      Cost in Dollars    Subject  
2117        2424.7344       AIDS  
2118        2424.7344       AIDS  
2119        1816.0560       AIDS  
2120        2424.7344       AIDS  
2121        2652.7380       AIDS  
2122        1761.0780  undefined  
2123        2421.8964       AIDS  
2124        2421.8964  undefined  
2125        3134.3664       AIDS  
2126        2685.8700  undefined

In [81]:
df["Subject"].value_counts()

undefined     1639
chemistry      168
biology        150
psychology      82
immunology      46
physiology      23
AIDS            19
Name: Subject, dtype: int64

In [84]:
df[df['Subject']=='undefined']['Journal title'].value_counts()

plos one                                                      199
neuroimage                                                     29
plos genetics                                                  24
plos pathogens                                                 24
nucleic acids research                                         23
proceedings of the national academy of sciences                20
plos neglected tropical diseases                               20
human molecular genetics                                       19
nature communications                                          19
bmc public health                                              14
brain                                                          14
movement disorders                                             13
developmental cell                                             12
journal of neurosciences                                       12
journal of general virology                                    11
bmj       

In [62]:
df[df["Subject"] == "undefined"]

PMID/PMCID  \
1                                            PMC3679557   
6                                            PMC3709265   
7                                  23057412 PMC3495574    
15                                           PMC3572711   
23                                       PMCID: 3584654   
32                                           PMC3411196   
34                                             23373658   
35                                           PMC3673662   
45                                 23025831 PMC3462035    
46                                 23731076 PMC3670620    
47                                  23244291 PMC3569044   
48                                           PMC3444304   
59                              PMID: 23409903 23409903   
81                                              3947303   
87                                           PMC3697472   
88                                           PMC3571300   
89                            PMID:22988015 PMC3497410    
90                                      PMC in progress   
91                                    PMCID: PMC3497412   
92                                           PMC3754575   
93                                                  NaN   
94                            PMID: 23152534 PMC3554137   
95                             PMCID:\n    PMC3648151\n   
96                           PMCID:\n    PMC3518918\n\n   
99    PMCID:\n    PMC3554100\n    [Available on 2013...   
100                                                 NaN   
101                                                 NaN   
102                                                 NaN   
103                                23255554 PMC3653566    
104                                                 NaN   
...                                                 ...   
2071                         PMID: 23627368  PMC3806363   
2074                                         PMC3569874   
2075                                         PMC3842118   
2076                                      PMID:23080298   
2077                                         PMC3867604   
2078                                         PMC3506972   
2079                                         PMC3731578   
2080                                      PMID:23696415   
2082                                           23837793   
2085                                           23386329   
2086                                    PMID: 22360292    
2087                                   PMCID:PMC3708113   
2088                                            3838626   
2089                             PMCID:\n    PMC3736148   
2092                                  Not yet available   
2098                         PMID: 22188548 PMC3492912    
2099                         PMID: 22764734 PMC3798095    
2100                                     PMID: 23703895   
2101                                  PMCID: PMC3664106   
2102                                     PMID: 23649934   
2103                                         PMC2955965   
2104                         PMID: 19948006 PMC3551259    
2105                                     PMID: 23305527   
2106                                  PMCID: PMC3600532   
2107                                         PMC3440570   
2113                          PMID:22997008 PMC3569869    
2114                                            3708104   
2122                                            2901593   
2124                                            3785148   
2126                   PMID: 23846567  (Epub July 2013)   

                                         Publisher  \
1                                              ACS   
6                                              ACS   
7                                              ACS   
15                                   AGA Institute   
23                       AMERICAN CHEMICAL SOCIETY   
32                       American Chemical Society   
34      

In [135]:
df.groupby("Subject").describe()

Article title         \
                   count unique   
Subject                           
AIDS                  19     19   
biology              139    139   
chemistry            168    168   
immunology            33     33   
physiology            23     23   
psychology            82     82   
undefined           1663   1662   

                                                                    \
                                                          top freq   
Subject                                                              
AIDS        HIV incidence and survival from age-specific s...    1   
biology     An experimental model for studying the biomech...    1   
chemistry   Regioselective opening of myo-inositol orthoes...    1   
immunology  Rapid early innate control of hepatitis C viru...    1   
physiology  Bimodal collagen fibril diameter distributions...    1   
psychology  COMT Val158Met Genotyp Determines the Directio...    1   
undefined   Exclusive breastfeeding, diarrhoel morbidity a...    2   

           COST (£) charged to Wellcome (inc VAT when charged)         \
                                                         count unique   
Subject                                                                 
AIDS                                                       19      15   
biology                                                   139     117   
chemistry                                                 168     144   
immunology                                                 33      31   
physiology                                                 23      23   
psychology                                                 82      58   
undefined                                                1663    1088   

                          Journal title         \
                 top freq         count unique   
Subject                                          
AIDS        £1836.92    3            19     10   
biology     £1939.95    6           139     61   
chemistry   £1800.00   13           168     60   
immunology  £3000.00    3            33     14   
physiology  £2110.15    1            23     11   
psychology  £3000.00    9            82     46   
undefined   £2040.00   84          1662    726   

                                                 PMID/PMCID         \
                                        top freq      count unique   
Subject                                                              
AIDS                              aids care    4         19     19   
biology                     current biology   11        123    123   
chemistry   journal of biological chemistry   53        164    163   
immunology   european journal of immunology    8         32     32   
physiology            journal of physiology    8         20     20   
psychology                 neuropsychologia    7         73     70   
undefined                          plos one  190       1497   1461   

                                          Publisher         \
                                 top freq     count unique   
Subject                                                      
AIDS        PMCID:\n    PMC3564972\n    1        19     10   
biology                   PMC3477309    1       139     41   
chemistry                 PMC3173209    2       168     45   
immunology         PMCID: PMC3298641    1        33     16   
physiology                   3764640    1        23     11   
psychology                PMC3529206    2        82     24   
undefined          Not yet available    7      1663    238   

                                                 
                                       top freq  
Subject                                          
AIDS                        Wolters Kluwer    7  
biology                           Elsevier   49  
chemistry                            ASBMB   44  
immunology                        Elsevier    9  
physiology  American Physiological Society    7  
psych

The above table is the closest I have come to finding the open access prices by subject area.